### PR TITLE
cli: Provide better flags for the "less" pager than default

### PR DESCRIPTION
**Pull request by benbrittain**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (cli/src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/config.rs [hunk 1]
```diff
@@ -324,7 +324,17 @@ fn env_base() -> config::Config {
...
     if let Ok(value) = env::var("PAGER") {
-        builder = builder.set_override("ui.pager", value).unwrap();
+        if value == "less" {
+            // Many systems ship with "R" as the only setting in $LESS, which provides
+            // a less than ideal experience with jj, since they won't see the log history
+            // persist on the screen. While slightly inconsistent to ignore that setting,
+            // it should provide a better "out-of-box experience". If a user truly wants
+            // to change the values that are passed to "less", they can set that in the
+            // ui.pager jj config.
+            builder = builder.set_override("ui.pager", "less -FRX").unwrap();
```
Comment (martinvonz) [inline cli/src/config.rs:334]: Maybe we should simply not do anything here and inherit the default value instead:
https://github.com/martinvonz/jj/blob/bcdb9beb6ce5ba625ae73d4839e4574db3d9e559/cli/src/config/misc.toml#L9

Comment (benbrittain) [reply cli/src/config.rs]: Seems fine to me, basically if your pager is set to less, you get the default settings instead :+1:

Comment (martinvonz) [reply cli/src/config.rs]: Yes, and we don't need to duplicate the `LESSCHARSET` bit. OTOH, if we later change the default to something else, you'll get that new default if you say `PAGER=less`. Hopefully we'll notice that :)

**Review by martinvonz — APPROVED** [review]
LGTM, but it's probably good to give others a day or something to spot any problems we've missed.

**Comment by ilyagr** [comment]
I think this should be documented in https://martinvonz.github.io/jj/v0.13.0/config/#pager. The simplest thing would be to change a sentence to say:

> less -FRX is the default pager in the absence of any other setting, or if `$PAGER` is exactly equal to `less`.

You could also add footnote about how to disable this behavior.

If it's a chore, this doesn't have to be a blocker; we can do it in a separate PR.

Overall, I think this could work; we'll see if anybody gets annoyed by it. I can imagine somebody not liking `-FX`, but I think these people will be few.

**Comment by yuja** [comment]
> Overall, I think this idea could work; we'll see if anybody gets annoyed by it. I can imagine somebody not liking `-FX`, but I think these people will be few.

I don't like `-F`, but I'm not strongly against this change. User can always override the default.

fwiw, both git and hg respect `$LESS` if set.

**Comment by benbrittain** [comment]
I almost feel like the actual right answer is to have a built-in pager with the defaults `jj` would prefer, then allow users to set preferring the system pager if they want.

**Comment by ilyagr** [comment]
> I almost feel like the actual right answer is to have a built-in pager with the defaults `jj` would prefer, then allow users to set preferring the system pager if they want.

We discussed this a few times, but unfortunately didn't create a bug for it.

https://github.com/markbt/streampager would be a nice option, it's what Sapling uses. It seems no longer maintained. I think I found a few more Rust pagers, but they seemed very new. There are some nice Golang pagers (`moar` and `ov`), but I don't know how to link them to a Rust program. (They are not perfect; I tried using them instead of `less` for a while, but there is always some minor issue that made me switch back to `less`. But they have better defaults).

**Comment by jyn514** [comment]
> There are some nice Golang pagers (`moar` and `ov`), but I don't know how to link them to a Rust program.

as a last resort, if you really want to distribute jj as a single static binary, you could embed the entire Go binary as static data in the rust binary, and then unpack it at runtime into a binary in the same directory (or run it in-memory with https://docs.rs/memfd-exec). but i know that's kinda hacky.

File: cli/src/config.rs [hunk 2]
```diff
@@ -324,7 +324,16 @@ fn env_base() -> config::Config {
...
+        // If 'less' is the value of PAGER, do not override the pager.
+        // Many systems ship with "R" as the only setting in $LESS, which provides
+        // a less than ideal experience for 'less' with jj, since they won't see the
+        // log history persist on the screen. While slightly inconsistent to ignore
+        // that setting, it should provide a better "out-of-box experience". If a user
+        // truly wants to change the values that are passed to "less", they can set that
+        // in the ui.pager jj config.
+        if value != "less" {
+            builder = builder.set_override("ui.pager", value).unwrap();
+        }
```
Comment (martinvonz) [inline cli/src/config.rs:336]: Maybe this change should be mentioned in the changelog?

**Comment by epage** [comment]
As I commented on the Issue, if you are interesting all of what git does, see https://github.com/epage/git-dive/blob/main/src/git_pager.rs

As for built-in pagers, one I hadn't seen mentioned is https://crates.io/crates/minus.

**Comment by ilyagr** [comment]
Right, that's a good one! `procs` uses `minus` in [exactly the way we would](https://github.com/dalance/procs/tree/master?tab=readme-ov-file#pager), so we can learn from them.

**Comment by benbrittain** [comment]
Closing in favor of #3024 and anything else that will come from that direction of a solution instead of special casing `less`

**Pull request closed without merging**